yacht.sh

```diff
@@ -0,0 +1,187 @@
+  error_code=$?
+  popd &>/dev/null || exit 1
+
+  # propagate the error out if an error code occured.
+  if [ $error_code -ne 0 ]; then
+    exit $error_code
+  fi
+}
+
+
+
+
+build_type=RelWithDebInfo
+be_verbose="false"
+do_format="false"
+
+while test $# -gt 0; do
+  case "$1" in
+  -h | --help)
+    echo "build.sh - Builds the source with cmake and make."
+    echo " "
+    echo "./build.sh [-h|--help] [-r|--release] [-d|--debug] [-c|--clean] TARGET"
+    echo " "
+    echo "options:"
+    echo "  -h, --help                show brief help"
+    echo "  -r, --release             use the Release build type    (mutually exclusive with -d,--debug,-m, and --minsizerel)"
+    echo "  -d, --debug               use the Debug build type      (mutually exclusive with -r,--release,-m, and --minsizerel)"
+    echo "  -m, --minsizerel          use the MinSizeRel build type (mutually exclusive with -r,--release,-d, and --debug)"
+    echo "  -c, --clean               cleans the build directory before compiling"
+    echo "  TARGET                    specify a compile target"
+    echo "  -v, --verbose             use verbose output"
+    echo " "
+    echo "if neither -r,--release,-d,--debug,-m, or --minsizerel are set, then the default build type used is Relwithdebinfo."
+    exit 0
+    ;;
+  -f | --format)
+    do_format="true"
+    shift
+    ;;
+  -v | --verbose)
+    be_verbose="true"
+    shift
+    ;;
+  -r | --release)
+    build_type=Release
+    shift
+    ;;
+  -d | --debug)
+    build_type=Debug
+    shift
+    ;;
+  -m | --minsizerel)
+    build_type=MinSizeRel
+    shift
+    ;;
+  -c | --clean)
+    clean=ON
+    shift
+    ;;
+  *)
+    target=$1
+    shift
+    ;;
+  esac
+done
+
+if [ -n "$clean" ]; then
+  compile ${build_type} clean
+  echo "Cleaned build directory."
+  echo " "
+fi
+
+compile ${build_type} "${target}"
+
+if [ "$do_format" = "true" ]; then
+
+  format_command="clang-format -i -style=file %"
+  if [ "$be_verbose" = "true" ]; then
+    verbose_command="clang-format --dry-run --Werror --verbose -style=file %"
+    format_command="${verbose_command} ; ${format_command}"
+  fi
+
+  if ! hash clang-format 2>/dev/null; then
+    echo "'clang-format' is not installed on this machine, cannot use it to format code. Halting..."
+    if hash apt 2>/dev/null; then # if apt is install, give a suggestion about how to install clang-format
+      echo "It can be installed with:"
+      echo ""
+      echo "sudo apt install clang-format"
+      echo ""
+    else
+      echo "It can install it from LLVM's tools:"
+      echo ""
+      echo "https://releases.llvm.org/download.html"
+      echo ""
+    fi
+    exit 1
+  fi
+
+  for path in "./src/" "./include/" "./libs/"; do
+    if [ -d "${path}" ]; then
+      find "${path}" \( -name "*.c" -o -name "*.cc" -o -name "*.cpp" -o -name "*.h" -o -name "*.hpp" -o -name "*.ipp" \) -print |
+        xargs -I % sh -c "${format_command}"
+    fi
+  done
+
+fi
```

Comment:
Format should be above `compile ${build_type} "${target}"`, in case some madlad wants to clean, format, & compile in one go... Otherwise it will want to recompile as the source files changed.